src/python/pants/engine/README.md

```diff
@@ -105,29 +111,32 @@ class TypedDatatype(datatype([('field_name', Exactly(str, int))])):
 ```
 
 Assigning a specific type to a field can be somewhat unidiomatic in Python, and may be unexpected or
-unnatural to use. Additionally, the engine already applies a form of implicit type checking by
-ensuring there is a unique path from subject to product when a product request is made. However,
-regardless of whether the object is created directly with type-checked fields or whether it's
-produced from a set of rules by the engine's dependency injection, it is extremely useful to
-formalize the assumptions made about the value of an object into a specific type, even if the type
-just wraps a single field. The `datatype()` function makes it simple and efficient to apply that
-strategy.
+unnatural to use. However, regardless of whether the object is created directly with type-checked
+fields or whether it's produced from a set of rules by the engine's dependency injection, it is
+extremely useful to formalize the assumptions made about the value of an object into a specific type,
+even if the type just wraps a single field. The `datatype()` function makes it simple and efficient
+to apply that strategy.
+
```

Comment:
Maybe outside the scope of your PR, but would be useful to enumerate all the type constraints like `SubclassOf`.